[DISCO-2625] Migrate Top Picks File Processing to GCS

taddes opened the pull request:
## References

JIRA: [DISCO-2625](https://mozilla-hub.atlassian.net/browse/DISCO-2625)

## Description
Migrate Top Picks File Processing to GCS



## PR Review Checklist

_Put an `x` in the boxes that apply_

- [ ] This PR conforms to the [Contribution Guidelines](https://github.com/mozilla-services/merino-py/blob/main/CONTRIBUTING.md)
- [ ] The PR title starts with the JIRA issue reference, format `[DISCO-####]`, and has the same title (if applicable)
- [ ] `[load test: (abort|warn)]` keywords are applied (if applicable)
- [ ] [Documentation](https://github.com/mozilla-services/merino-py/tree/main/docs) has been updated (if applicable)
- [ ] [Functional and performance test](https://github.com/mozilla-services/merino-py/blob/main/docs/dev/testing.md) coverage has been expanded and maintained (if applicable)

ncloudioj commented on merino/providers/top_picks/backends/top_picks.py:
```diff
@@ -1,18 +1,129 @@
+        """
+        try:
+            bucket: Bucket = client.get_bucket(self.gcs_bucket_path)
+            domain_files: Any = bucket.list_blobs(delimiter="/")
```
Just noticed that there is still a timestamp in the file name for the latest file in the bucket, which makes it harder to use over here. I'd suggest changing the naming convention in the bucket for the latest file, namely from "{timestamp}_top_picks_latest.json" to "top_picks_latest.json" so that we don't need to call `list_blobs()` here, we can do this instead:

```py
blob: Optoinal[Blob] = bucket.get_blob(
    "top_picks_latest.json",
    if_generation_not_match = self.blob_generation  // assuming we store the `generation` upon the last fetch 
)
```

This would only return a `Blob` if the provided generation doesn't match that of the file in the bucket, meaning that the latest file has been updated. Otherwise, it returns `None`, meaning that the file remains the same, we don't need to do anything.

ncloudioj commented on merino/providers/top_picks/backends/top_picks.py:
```diff
@@ -1,18 +1,129 @@
+            ) from json_error
+
+    @staticmethod
+    def _parse_date(file_path: str) -> datetime | None:
```
Perhaps we don't need this helper function for the local file manager, correct?

taddes commented on merino/configs/default.toml:
```diff
@@ -365,6 +365,29 @@ firefox_char_limit = 2
+
+# MERINO_PROVIDERS__TOP_PICKS__GCS_PROJECT
+# GCS project name that contains domain data
+gcs_project = "moz-fx-merino-prod-1c2f"
```
Sounds good

taddes commented on merino/configs/default.toml:
```diff
@@ -287,7 +287,7 @@ backend = "remote-settings"
 # The cron job should tick more frequently than `resync_interval_sec` so that
 # the resync failure can be retried soon.
-cron_interval_sec = 60
+cron_interval_sec = 6000
```
Ack, my bad, this was meant for the `cron_interval_sec` under the Top Pick Provider, not ADM. Good catch!

ncloudioj commented on merino/providers/top_picks/backends/top_picks.py:
```diff
@@ -1,21 +1,132 @@
+            current_date: datetime = datetime.now()
+            file_contents: dict = json.loads(blob_data)
+            logger.info(
+                f"Domain file {blob.name} acquired. File generated on: {blob_date}."
```
I don't think we'd care too much about those timestamps, rather we can log something like "Succeeded to load the remote domain file" when the whole indices get built successfully. Also, we want to log and record metrics for those exceptions.

taddes commented on merino/providers/top_picks/backends/top_picks.py:
```diff
@@ -1,21 +1,132 @@
+        """
+        try:
+            bucket: Bucket = client.get_bucket(self.gcs_bucket_path)
+            blob: Blob = bucket.get_blob("top_picks_latest.json")
```
Makes sense @ncloudioj ! And my bad, I originally thought `if_generation_not_match` was just an alias term for a function I'd add that returns a bool, then it hit me it's a param passed in that is a function in the bucket object that does all those cool things for us... sorry for that confusion!

taddes commented:
Hey @ncloudioj ! Made the requested changes and now have that inbuilt GCS check regarding generation. I had to refactor some things accordingly to support the return value of `None` should there be no file to capture. Added a few mypy ignore statements in tests and such due to it not being smart enough to intuit the diff, but we're covered in the tests.

Also extracted the filemanager out into its own module for isolation, ease of use and testing purposes. Just seemed sensible. 

SRE working on svc account and IAM permissions for us in the meantime

taddes commented on merino/providers/top_picks/backends/top_picks.py:
```diff
@@ -135,8 +149,36 @@ def build_index(self, domain_list: dict[str, Any]) -> TopPicksData:
 
-    def build_indices(self) -> TopPicksData:
+    def build_indices(self) -> TopPicksData | None:
         """Read domain file, create indices and suggestions"""
```
Great suggestion!

taddes commented on merino/providers/top_picks/provider.py:
```diff
@@ -23,24 +29,67 @@ class Provider(BaseProvider):
+                "Failed to fetch data from Top Picks Backend.",
+                extra={"error message": f"{backend_error}"},
+            )
+            raise BackendError
```
Makes sense, I also am removing the `raise BackendError` from its instance below in `_fetch_top_picks_data`

taddes commented on merino/providers/top_picks/backends/filemanager.py:
```diff
@@ -0,0 +1,99 @@
+        self.gcs_bucket_path = gcs_bucket_path
+        self.blob_generation = 0
+
+    def create_gcs_client(self) -> Client:
```
Ah, so this was done this way for improving ease of testing and modularity.  Was a suggestion encouraged by Kat. Unless we can inject the client, patching gets really gnarly and messy. Returning a client this way and then passing to `get_file` means fewer headaches.

taddes commented:
Hey @ncloudioj ! Refactored this to utilize the enum pattern we discussed. Since the filemanager is where the generation value is checked, determining whether there is a skip, pass or fail, it's passed onto the backend and then to provider, which sets the `last_fetch_at` only in the case of `SUCCESS`. Let me know how this looks to you, all other previous changes we discussed were added prior to this last commit. Thx for the idea!

ncloudioj commented on merino/providers/top_picks/backends/filemanager.py:
```diff
@@ -0,0 +1,107 @@
+                blob_data = blob.download_as_text()
+                file_contents: dict = json.loads(blob_data)
+                logger.info("Successfully loaded remote domain file.")
+                return (GetFileResultCode.SUCCESS, file_contents)
```
Nice!

ncloudioj reviewed (approved):
This looks great, thanks for making this happen!
